core/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala

```diff
@@ -91,6 +92,51 @@ private[spark] abstract class YarnSchedulerBackend(
   }
 
   /**
+   * Override the DriverEndpoint to add extra logic for the case when an executor is disconnected.
+   * We should check the cluster manager and find if the loss of the executor was caused by YARN
+   * force killing it due to preemption.
+   */
+  private class YarnDriverEndpoint(rpcEnv: RpcEnv, sparkProperties: ArrayBuffer[(String, String)])
```

Comment:
Wait, I'm confused. `YarnSchedulerEndpoint` and `DriverEndpoint` are different things; `YarnSchedulerEndpoint` is a communication channel between the driver in YARN mode and the YARN AM, there are no executors involved. Why wouldn't that work here?